### ISSUE TITLE
FilePath should allow access to all possible paths on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2366]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#2366|
|Type|defect|
|Created|2007-01-07 04:38:24Z|

`FilePath` incorrectly prohibits access to certain files, due to the crummy APIs it's using.

For example, you can actually have a file called "CON" on windows, and it is accessible by quoting the path in various ways (or by accessing it via native Windows APIs), but `FilePath` would currently consider that path a security violation, regardless of whether it actually exists.

Additionally, Windows paths may natively be unicode.  `FilePath` doesn't really pay attention to unicode vs str, so it may be broken with respect to non-ASCII(?) paths.  Even if it does work, it needs unit tests to demonstrate this and prevent regressions.

Attachments:

* [twisted.python.filepath-2366-1.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/dab/daba40f9fd3019a1f8eeb5dab88a784c59fbb794/0ecf3b7b968c1e60c1baac173c90a17290e82a87.patch) (53842 bytes) - added by hbock on 2013-07-14 20:32:59Z - first stab at a fix for both Python 2.6+/3.3+

<details><summary>Searchable metadata</summary>

```
trac-id__2366 2366
type__defect defect
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__core core
keywords__unicode_windows unicode windows
time__1168144704000000 1168144704000000
changetime__1441823005825629 1441823005825629
version__None None
owner__ 
cc__teratorn cc__Zancas cc__davidsarah cc__zooko@... cc__hbock
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_2"></a>|@glyph commented|
|-|-|

This should also include using ```MOVE_FILE_REPLACE_EXISTING``` to actually make ```setContent``` (and ``` moveTo```, of course) atomic.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

I have created [#3454](https://github.com/twisted/twisted/issues/3454) for the separate issue of `setContent` and `moveTo` being non-atomic.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="Zancas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|Zancas commented|
|-|-|

Replying to [<automation>](#note_6):
 Hi all I see that [#4736](https://github.com/twisted/twisted/issues/4736) was closed as a duplicate of this ticket, but it's not clear to me that FilePath handles Unicode in all cases.  Does it?

**Comment by twisted-trac** [comment]
|<img alt="Zancas's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|Zancas commented|
|-|-|

Replying to [<automation>](#note_6):
 Hi all I see that [#4736](https://github.com/twisted/twisted/issues/4736) was closed as a duplicate of this ticket, but it's not clear to me that FilePath handles Unicode in all cases.  Does it?

**Comment by twisted-trac** [comment]
|<img alt="davidsarah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|davidsarah commented|
|-|-|

The Tahoe-LAFS project [would like to use FilePath](http://tahoe-lafs.org/trac/tahoe-lafs/ticket/1437), but we're concerned that lack of Unicode support would cause a regression on Windows relative to our existing code that uses ```unicode``` strings for paths.

**Comment by twisted-trac** [comment]
|<img alt="davidsarah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|davidsarah commented|
|-|-|

By the way, I don't really agree that it is a good idea to create Windows files with names that are usually prohibited. The Microsoft article on this is pretty clear:
```
Do not use the following reserved device names for the name of a file:

CON, PRN, AUX, NUL, COM1, COM2, COM3, COM4, COM5, COM6, COM7, COM8, COM9, LPT1, LPT2, LPT3, LPT4, LPT5, LPT6, LPT7, LPT8, and LPT9. Also avoid these names followed immediately by an extension; for example, NUL.txt is not recommended. For more information, see Namespaces.
```
The fact that you _can_ access these files via ```\\?\``` is beside the point; you'll end up with files that can't be accessed by most other applications.

But [[#4736](https://github.com/twisted/twisted/issues/4736)](https://github.com/twisted/twisted/issues/4736) was closed as a duplicate of this ticket. Perhaps #4736 should be reopened in order to focus on the Unicode issue.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

[#4736](https://github.com/twisted/twisted/issues/4736) was just a bad ticket.  I would have closed it as invalid instead of as a duplicate.

If someone wants to work on the issue of unicode paths separately from the other issues described on this ticket, file a new ticket for it.  As [James pointed out](http://twistedmatrix.com/trac/ticket/4736#comment:6) though, there is some work in the branch associated with this ticket which addresses unicode issues.

**Comment by twisted-trac** [comment]
|[<img alt="zooko's avatar" src="https://avatars.githubusercontent.com/u/467900?s=50" width="50" height="50">](https://github.com/zooko)<a name="note_13"></a>|@zooko commented|
|-|-|

I created [#5203](https://github.com/twisted/twisted/issues/5203) (FilePath.children() should return FilePath objects with unicodes in them instead of strs) to track a specific feature that we could use. Also to publish a small novel about Tahoe-LAFS and filename handling.

**Comment by twisted-trac** [comment]
|<img alt="davidsarah's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|davidsarah commented|
|-|-|

Replying to [exarkun](#note_12):
> [#4736](https://github.com/twisted/twisted/issues/4736) was just a bad ticket.  I would have closed it as invalid instead of as a duplicate.
> 
> If someone wants to work on the issue of unicode paths separately from the other issues described on this ticket, file a new ticket for it.  As [James pointed out](http://twistedmatrix.com/trac/ticket/4736#comment:6) though, there is some work in the branch associated with this ticket which addresses unicode issues.

Er, what branch? I don't see anything linked from the Branch field. Am I looking in the wrong place?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_15"></a>|@glyph commented|
|-|-|

Replying to [davidsarah](#note_14):
> Replying to [exarkun](#note_12):
> > [#4736](https://github.com/twisted/twisted/issues/4736) was just a bad ticket.  I would have closed it as invalid instead of as a duplicate.
> > 
> > If someone wants to work on the issue of unicode paths separately from the other issues described on this ticket, file a new ticket for it.  As [James pointed out](http://twistedmatrix.com/trac/ticket/4736#comment:6) though, there is some work in the branch associated with this ticket which addresses unicode issues.
> 
> Er, what branch? I don't see anything linked from the Branch field. Am I looking in the wrong place?

Some branches predate the auto-branch-field-updater gizmo so you have to go to http://twistedmatrix.com/trac/browser/branches/ and grovel around for the ticket number.  I've updated the branch field on this ticket.

**Comment by twisted-trac** [comment]
|<img alt="hbock's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|hbock commented|
|-|-|

I've attached a first stab at a patch that resolves this issue, per the advice of Itamar on the mailing list.

Sorry the patch is so large.  Most of the changes involve getting rid of the b"" qualifier for byte strings in the unit tests. 

 - Use os.O_BINARY instead of importing it from the win32 module - supported in 2.6+
 - Rename _secureEnoughString to _secureEnoughBytes to reflect usage
 - Remove old pre-2.6 exception handling for listdir() on Windows - this no longer applies and can be handled with one except clause
 - listdir() is expected to return a list in other modules (trial); map() returns a view on Python 3.3, use list comprehensions instead
     - This also applies to globChildren()
 - Add __eq__ support
 - Update documentation to explain str/bytes behavior and restrictions
 - Depending on whether the input pathname string is unicode or bytes, internally cache certain constants as the right type so Python won't raise exceptions for various operators and system calls.
 - setContent and temporarySibling now use None as a placeholder for their default extensions, so we can properly infer the correct string type to use without needing to decode/encode
 - Change old unittests back to using the native string type, which should continue to work as expected for both Python 2.6+ and 3.3+.
 - Add unittests that explicitly deal with unicode vs. bytes handling.

One question I have is regarding the suppression of warnings; I force Python to filter the ntpath DeprecationWarnings on Windows during unit tests. Is this acceptable?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

Thanks for your work on this.  Just one comment for starters, sorry for not doing a complete review instead.

> Most of the changes involve getting rid of the b"" qualifier for byte strings in the unit tests.

I don't think this is the right approach.  All strings should either be b"" or u"", both of which are legal in both Python 2.{6,7} and Python 3.3, the three versions of CPython Twisted targets at this time.

Using "" means the test has different meaning depending on which version of Python is being used.  Using b"" or u"" means the test has the same meaning no matter what.  We should aim to retain the current behavior for byte strings, tested on Python 2.x and Python 3.x.  We'll presumably also need new tests that verify the desired behavior for unicode strings has been implemented.

[Plan/Python3](https://twistedmatrix.com/trac/wiki/Plan/Python3) tries to explain this idea along with some others.

**Comment by twisted-trac** [comment]
|<img alt="hbock's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_19"></a>|hbock commented|
|-|-|

Removed from review per the suggestion of Tom Prince to break it up.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

This ticket talks about "native APIs", implying that `CreateFile` doesn't have the same restrictions on CON and NUL that C stdio `open` does.

I've experimented with it though via pywin32, and it looks like it does.

This documentation suggests this feature can't be disabled, and doesn't give any guidance on quoting the path name - https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx#consoles

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl commented|
|-|-|

I believe that this is fixed, by way of [#7805](https://github.com/twisted/twisted/issues/7805). Think so, Glyph?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph set status to `closed`|
|-|-|

That's... probably as close as we're going to get.  I still have no idea if it's possible to create a file called "CON" :).
